doc/SAI-Proposal-SAI-PTF.md

```diff
@@ -0,0 +1,132 @@
+- **sai_adapter.py** - python wrappers for calling SAI functions
+
+These files should be generated once, per every SAI version that needs to be tested.
+
+| Input files                                                  | Output files                                                 |
+| ------------------------------------------------------------ | ------------------------------------------------------------ |
+| **Templates** <br /> - *sai.thrift.tt* <br /><br />**SAI headers** <br /> - *SAI meta* generates XML files with *Doxygen* | **sai.thrit** <br /> - thrift RPC "header" <br /> - generated from scratch <br /> - needed by thrift to generate *sai_rcp_server.skeleton.cpp* <br /> - skeleton is used to generate *sai_rpc_server.cpp.tt* template |
+| **Templates** <br /> - *sai_rpc_server_functions.tt* <br /> - *sai_rpc_server_helper_functions.tt* <br /><br />**Auto-generated templates** <br /> - *sai_rpc_server.cpp.tt* (by Thrift from *sai.thift*) <br /><br />**SAI headers** <br /> - *SAI meta* generates XML files with *Doxygen* | **sai_rpc_server.cpp** <br /> - thrift server <br /> - need *sai.thrift* and *Thrift* itself for generation |
+| **Templates** <br /> - *sai_adapter.py.tt* <br /> - *sai_adapter_utils.tt* <br /><br />**SAI headers** <br /> - *SAI meta* generates XML files with *Doxygen* | **sai_adapter.py** <br /> - thrift client and main SAI PTF test library <br /> - generated from scratch |
+
+
+All of this can be generated with gensairpc.pl script.
+
+Autogeneration:
+
+```
+cd <root>/sai/
+./gensairpc.pl --clean
+```
+
+When upgrading SAI:
+
+```
+cd SAI/meta
+make
+cp saimetadata.c <root>/sai
+cp saimetadata.h <root>/sai
+cd <root>/sai
+./gensairpc.pl --clean --no-meta-build
+make clean - C SAI/meta
+```
+
+Use --no-meta-build because SAI meta are already built with make.
+
+## Test execution
+
+Test files written in Python call functions from sai_adapter.py. When running such test auto-generated Python functions are talking with C-based RPC server to run C functions. These functions are used to create/remove SAI object and get/set their attributes. Then it is running internal SAI implementation adopting common SAI interface to test the implementation.
+
+Tests use SAI to configure underlay switch. Using PTF we send and verify packets that are testing each switch configuration.
+
+## Tests structure
+
+The tests themselves are based on python unittest framework (https://docs.python.org/3/library/unittest.html). They consist of three main functions:
+
+- setUp() - creating necessary     objects
+- runTest() - running proper test
+- tearDown() - removing all     created objects
+
+Tests have to inherit from one of two base classes - SaiHelper or SaiHelperBase. SaiHelperBase is the base class that runs initial configuration for all tests. SaiHelper inherits from SaiHelperBase but provides additional configuration.
+
+The based configuration created by SaiHelper looks like this:
+
```

Comment:
We can also add one example SAI object and list the autogenerated APIs and how it is being invoked from PTF script